File: python/3108. Minimum Cost Walk in Weighted Graph.py
class Solution:
    def minimumCost(self, n: int, edges: List[List[int]], query: List[List[int]]) -> List[int]:
        uf = UnionFind(n)
        for (x, y, weight) in edges:
            uf.union(x, y, weight)
        answer = []
        for (x, y) in query:
            if x == y:
                answer.append(0)
            elif uf.find(x) != uf.find(y):
                answer.append(-1)
            else:
                weight = uf.findWeight(x)
                answer.append(0) if weight == (1 << 30) - 1 else answer.append(weight)
        return answer
        
class UnionFind:
    def __init__(self, n):
        self.parent = list(range(n))
        self.bitwise_and = [(1 << 30) - 1] * n

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def findWeight(self, x):
        return self.bitwise_and[self.find(x)]

    def union(self, x, y, weight):
        root_x = self.find(x)
        root_y = self.find(y)
        if x < y:
            self.parent[root_y] = root_x
            self.bitwise_and[root_x] &= self.bitwise_and[root_y] & weight
        else:
            self.parent[root_x] = root_y
            self.bitwise_and[root_y] &= self.bitwise_and[root_x] & weight

File: java/3108. Minimum Cost Walk in Weighted Graph.java
class Solution {
    public int[] minimumCost(int n, int[][] edges, int[][] query) {
        UnionFind uf = new UnionFind(n);
        for (int[] edge : edges) {
            uf.union(edge[0], edge[1], edge[2]);
        }
        int[] answer = new int[query.length];
        for (int i = 0; i < query.length; i++) {
            int x = query[i][0];
            int y = query[i][1];
            if (x == y) {
                answer[i] = 0;
            } else if (uf.find(x) != uf.find(y)) {
                answer[i] = -1;
            } else {
                answer[i] = uf.findWeight(x);
                answer[i] = answer[i] == 2147483647 ? 0 : answer[i];
            }
        }
        return answer;
    }
}

class UnionFind {
    private int[] parent;
    private int[] bitwiseAnd;
    private static final int ONES = Integer.parseInt("1111111111111111111111111111111", 2);
  
    UnionFind(int n) {
        parent = new int[n];
        bitwiseAnd = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            bitwiseAnd[i] = ONES;
        }
    }

    int find(int x) {
        if (x != parent[x]) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    int findWeight(int x) {
        return bitwiseAnd[find(x)];
    }

    void union(int x, int y, int weight) {
        int rootX = find(x);
        int rootY = find(y);
        if (x < y) {
            parent[rootY] = rootX;
            bitwiseAnd[rootX] &= bitwiseAnd[rootY];
            bitwiseAnd[rootX] &= weight;
        } else {
            parent[rootX] = rootY;
            bitwiseAnd[rootY] &= bitwiseAnd[rootX];
            bitwiseAnd[rootY] &= weight;
        }
    }
}